src/runtime/security/kinit_context.cpp

```diff
@@ -0,0 +1,306 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "kinit_context.h"
+#include "utils/shared_io_service.h"
+
+#include <boost/asio/deadline_timer.hpp>
+#include <fmt/format.h>
+#include <krb5/krb5.h>
+
+#include <dsn/utility/defer.h>
+#include <dsn/utility/time_utils.h>
+#include <dsn/dist/fmt_logging.h>
+#include <dsn/utility/flags.h>
+#include <dsn/utility/filesystem.h>
+#include <dsn/utility/smart_pointers.h>
+
+namespace dsn {
+namespace security {
+extern bool FLAGS_enable_auth;
+
+#define KRB5_RETURN_NOT_OK(err, msg)                                                               \
+    do {                                                                                           \
+        krb5_error_code __err_code__ = (err);                                                      \
+        if (__err_code__ != 0) {                                                                   \
+            return krb5_call_to_errors(__err_code__, (msg));                                       \
+        }                                                                                          \
+    } while (0);
+
+DSN_DEFINE_string("security", krb5_keytab, "", "absolute path of keytab file");
+DSN_DEFINE_string("security", krb5_config, "", "absolute path of krb5_config file");
+DSN_DEFINE_string("security", krb5_principal, "", "kerberos principal");
+
+// Attention: we can't do these check work by `DSN_DEFINE_validator`, because somebody may don't
+// want to use security, so these configuration may not setted. In this situation, these checks
+// will not pass.
+error_s check_configuration()
+{
+    dassert(FLAGS_enable_auth,
+            "There is no need to check configuration if FLAGS_enable_auth is not true");
+
```

Comment:
The comment shows that the code try to get keytab files from KDC, why should we check file exists on disk? Do we download the files from KDC or just cache those information in memory?

Reply:
No, keytab file is provided in our service's local machine. The code get from KDC is TGT